### PR TITLE
fix(mirror): unstake after staking pool contracts add more validators

### PR DESCRIPTION
**Pull request by marcelo-gonzalez**
when we generate txs to send, we skip over any stake actions, because
we don't want to add validators that don't actually exist. But this
does nothing to stop extra validators from showing up after staking
pool contracts get called, and this is where lots of validators on
mainnet generate stake receipts.

One way to get around this would be to check whether a given
FunctionCall tx eventually resulted in a stake action, and avoid
sending it if so. This adds a good bit of complexity though, since it
might happen after several cross contract calls, and it moves us
slightly further away from reproducing the source chain traffic
exactly

So here what we're going to do is send extra stake transactions to
unstake any unwanted validators if we see stake actions show up in
processed receipts (and we generated them, which we can tell by
checking whether the receipt's hash was stored in the mirror rocksDB).
This requires adding some logic to be able to add a transaction with
the next available nonce after we've generated many txs for that
signer (so we have to bump their nonces up by one each).  Also this
requires making sure that every account has a full access key (which
validators with staking pools do not, otherwise they wouldnt be
secure) so that we can actually send stake txs

### CODE REVIEW

**Comment by nikurt** [comment]
when you unstake, doesn't this change account balance and therefore impact the outcome of other mirrored function calls? :thinking:

File: tools/mirror/src/chain_tracker.rs [hunk 1]
```diff
@@ -757,17 +841,40 @@ impl TxTracker {
...
+                                ReceiptEnumView::Action { actions, .. } => {
+                                    // since this receipt was recorded in the DB, it means that this stake action
+                                    // resulted from one of our txs (not an external/manual stake action by the test operator),
+                                    // so we want to reverse it.
+                                    for a in actions {
+                                        if let ActionView::Stake { public_key, stake } = a {
+                                            if stake > 0 {
+                                                staked_accounts.insert(
+                                                    (
+                                                        outcome.receipt.receiver_id.clone(),
```
Comment (nikurt) [inline tools/mirror/src/chain_tracker.rs:855]: This indentation is deep.
(That's a neutral comment, feel free to keep).

File: tools/mirror/src/lib.rs [hunk 2]
```diff
@@ -1638,23 +1766,35 @@ impl<T: ChainAccess> TxMirror<T> {
...
+                    .await?;
             }
         }
 
         self.queue_txs(&mut tracker, target_head, false).await?;
 
-        self.main_loop(tracker, target_height, target_head).await
+        self.main_loop(tracker, target_height, target_head, source_hash).await
     }
 }
```
Comment (nikurt) [inline tools/mirror/src/lib.rs:1800]: And this file is long, 1800 lines.

**Review by nikurt — APPROVED** [review]
This system is getting too big. It's hard to get an understanding how it work by simply reading the code, which spans both Rust and Python.

File: tools/mirror/src/lib.rs (hunk 2 above)
Comment (marcelo-gonzalez) [reply tools/mirror/src/lib.rs:1800]: yea... this whole thing could use a refactor: https://github.com/near/nearcore/blob/67a527bf39534bfbef808f62e71c35eafb13ad0b/tools/mirror/src/chain_tracker.rs#L124

File: tools/mirror/src/chain_tracker.rs (hunk 1 above)
Comment (marcelo-gonzalez) [reply tools/mirror/src/chain_tracker.rs]: yea true, quite ugly. refactored it

**Pull request merged**